examples/ComplexFormula/del-1-2/Delete12_Raw.rvm

```diff
@@ -0,0 +1,115 @@
+package rvm;
+
+Delete12() {
+    /**
+     * This data structure represents a record that describes which user deletes what data at what time.
+     **/
+    public static class DeleteRecord {
+        public final long ts;
+        public final String user;
+        public final String deleteData;
+
+        public DeleteRecord(long ts, String user, String deleteData) {
+            this.ts = ts;
+            this.user = user;
+            this.deleteData = deleteData;
+        }
+
+        public boolean equals(DeleteRecord other) {
+            if (other == null)
+                return false;
+            if (this.ts != other.ts)
+                return false;
+
+            return (this.user.equals(other.user) &&
+                    this.deleteData.equals(other.deleteData));
+        }
+
+        public String print() {
+            return "@" + ts + " delete (" + user + ", db1, " + deleteData + ")\n";
+        }
+    }
+
+    //the suspiciousRecords is the list of suspicious records.
+    private static Set<DeleteRecord> suspiciousRecords = new HashSet<>();
+    private static HashMap<Long, Set<String>> cond2HasHopeDataList = new HashMap<>();
+    private static Set<String> db1InsertedData = new HashSet<>();
+    //db2InsertedData is eq to cond2 is not satisfiable
+    private static Set<String> db2InsertedData = new HashSet<>();
+
+    private void removeValidatedRecords(String data) {
+        List<DeleteRecord> removedRecords = new ArrayList<>();
+        for (DeleteRecord deleteRecord : suspiciousRecords) {
+            if (deleteRecord.deleteData.equals(data)) {
+                removedRecords.add(deleteRecord);
+            }
+        }
+        suspiciousRecords.removeAll(removedRecords);
+    }
+
+    public static void printAllViolations() {
+        for (DeleteRecord suspiciousRecord : suspiciousRecords) {
+            if (db2InsertedData.contains(suspiciousRecord.deleteData)) {
+                System.out.println("Violation: " + suspiciousRecord.print());
+            } else {
+                Set<String> hopefulDataList = cond2HasHopeDataList.get(suspiciousRecord.ts);
+                if (hopefulDataList == null)
+                    System.out.println("Violation: " + suspiciousRecord.print());
+
+                else if (hopefulDataList.contains(suspiciousRecord.deleteData)) {
+                    System.out.println("Satisfy the second condition of the disjunctive formula");
+                } else {
+                    System.out.println("Violation: " + suspiciousRecord.print());
+                }
+            }
+        }
+    }
+
+event delete (String user,String db,String p,String data) {boolean delete = true; boolean insert = false;}
```

Comment:
w.r.t the monitor variables, do you mean static variables at the top level of monitor class?
In your previous email, you said " So all you should do is to take the Java code after "raw:" and insert it right after where you insert the event handler code, for each event."
And you also pointed out that, in order to make generated code compile, the user can insert artificial variables at the beginning of the event action code.

In my opinion, the boolean flag which indicates which event happens should be available (able to be referred to) in the raw code anyway. So it can make users' life easier if we generate those flag variables automatally, so that users do not need to declare them manually and they can just assume the flags are available in the raw monitoring code and they can directly use them. But in your previous email, it seems that you do not like the static variables to be generated automatically,  so I came up with an idea of initializing the flags at the beginning of each event. The raw monitoring code will be appended to the end of every event action code, so the flags do not need to be static: their life-cycle can be just the event action method, and I can automatically generate the initialization of local flags for each event.